Update `devise-two-factor` to version 5.0.0

mjankowski opened the pull request:
As part of the Rails 7 and 7.1 upgrades, we initially thought we'd have to bump/migrate devise-two-factor (which moves off of `attr_encrypted` gem and onto native Rails encrypted feature) ... but it turned out there was some Rails 7 compatibility work done in both gems so this wasn't strictly needed for those updates, and we could keep the older code at the time.

This PR is the first of what is basically a two step migration process that they have documented very well: https://github.com/devise-two-factor/devise-two-factor/blob/main/UPGRADING.md#upgrading-from-4x-to-5x

To sum that up a bit:

- Phase 1 -- update the gem, add a new column that is going to use AR built-in encryption, including addition of either rails "credentials" or env vars (I've chosen that approach here, consistent with other features), add a method to user class to handle the legacy values from prior version (copied from update guide here), deploy all this
- Phase 2 -- use a rake task or console script or whatever to migrate the values, eventually add migrations and more code changes to remove the old env vars and columns, deploy again

Opening for general feedback, and migration strategy discussion. To the extent we want to actively migrate with a rake task or other approach (probably?) may also need documentation/upgrade writeup.

mjankowski committed:
Update devise-two-factor to version 5.0.0

mjankowski committed:
Add `users.otp_secret` migration column

mjankowski committed:
Add legacy support method for devise-two-factor upgrade

mjankowski committed:
Encrypted key values in env files

mjankowski committed:
Move legacy otp secret initializer to concern instead

mjankowski committed:
Update user spec otp secret check

mjankowski committed:
Annotate the new column

ClearlyClaire commented:
This seems sensible, but this will require involvement from server admins, properly setting up `ACTIVE_RECORD_ENCRYPTION_PRIMARY_KEY`, `ACTIVE_RECORD_ENCRYPTION_DETERMINISTIC_KEY` and `ACTIVE_RECORD_ENCRYPTION_KEY_DERIVATION_SALT` in addition to the existing `SECRET_KEY_BASE` and instead of (well, in addition to, for the transition period) `OTP_SECRET`.

I also wonder if those shouldn't be defined in `config/secrets.yml` rather than scattered across multiple files.

ThisIsMissEm commented:
The line I was wondering for when I saw this PR was this:

> This gem will fall back to reading the OTP secret from the legacy columns if it cannot find one in the new otp_secret column. When you write a new OTP secret it will always be written to the new otp_secret column.

From the upgrade guide linked.

mjankowski commented:
Yeah - I think that's basically the migration path so you don't break previously established codes. Anyone who is newly established 2FA or updating existing will populate the new field, but anyone just trying to auth against their prior established code will hit the legacy path and still read out from the prior fields.

Once deployed, there's a rake task (example on that upgrade guide) you can run which will proactively migrate everyone, and then when that's completed you can remove the rake task, along with the old fields and the legacy method support.

I think this is probably safe and their migration path is well thought out and should work -- my only hesitation here is that our spec coverage in this area is not perfect, and includes some stubbed out values around the OTP interaction specifically, because its often a random value that you have generate consistently in multiple places. May be worth taking a pass at creating a "does an OTP value created the old way successfully sign someone in with the legacy support" spec to increase confidence.

renchap commented:
I have not looked closely at this, but usually for this kind of database changes, we do it with a migration.

Could we somehow do it here, with a migration that creates the new column, then migrates all secrets to the new column?

The second migration can be in a post-deploy migration if we need to ensure that the new code is deployed before the data is changed.

Then in a future version we can drop the old column?

mjankowski commented:
> I have not looked closely at this, but usually for this kind of database changes, we do it with a migration.
> Could we somehow do it here, with a migration that creates the new column, then migrates all secrets to the new column?

Yes, the migration in this PR right now adds the new column. There is not (yet) in this PR anything to do the migration -- either a rake task or post deploy migration would work equally well here I think.

> Then in a future version we can drop the old column?

Yes.

I think the open questions here are:

- Do we want to introduce a credential management thing, or stick with env vars? I'd be inclined to stick with env vars ... or if we are going to move to credentials feature, do that as a separate effort ahead of this change so that we are reducing the places our mistakes might be coming from.
- Timing-wise or release-wise ... do we want the dynamic to be that there's an initial release where we add the column AND ALSO start migrating data (in which case this PR needs to add a rake task or post deploy migration), or do we want the initial release to just add the column and run in dual new/legacy mode for a while, with the idea that a future release (or releases) will do the migration and then once the migration is done, drop the column.

renchap commented:
We will need @ClearlyClaire to chime in on this, but here are my opinions

> * Do we want to introduce a credential management thing, or stick with env vars? I'd be inclined to stick with env vars ... or if we are going to move to credentials feature, do that as a separate effort ahead of this change so that we are reducing the places our mistakes might be coming from.

I would prefer that we first introduce support for AR encryption support, then in this PR switch the encryption to it.
I am not sure how we want to feed the AR encryption keys, and this will probably require specific upgrade steps and documentation, so better to be in a PR that does only this?

For env vars vs credentials, I do not think that credentials could work for us as those require a file to be committed into git, or baked into the container image. This is fine for apps that you deploy internaly, probably not for OSS apps like ours.

> * Timing-wise or release-wise ... do we want the dynamic to be that there's an initial release where we add the column AND ALSO start migrating data (in which case this PR needs to add a rake task or post deploy migration), or do we want the initial release to just add the column and run in dual new/legacy mode for a while, with the idea that a future release (or releases) will do the migration and then once the migration is done, drop the column.

I do not think this is a good idea to do it over 3 releases (add the column, then run in dual, then drop), because at our current rate it means doing it over one year. I do not see a big drawback in doing the new column + migration at once, this is what we did previously for the settings iirc?

renchap commented:
> I think this is not what you mean here, but just to be super nuanced here -- through framework updates alone, we do already have "support" for AR native encryption, we are just not using it for this feature (or any others, I don't think).

By support, I meant configuring it properly. If we do not set any of the keys, then we cant use the feature. So we need to ensure that people will generate those, and probably need to fail at startup if they are not defined, with the related docs on how to do so.

mjankowski committed:
Keep test env values in dot env file

mjankowski committed:
Data migration task

renchap commented on lib/tasks/devise_two_factor_migration.rake:
```diff
@@ -0,0 +1,17 @@
+  desc 'Copy devise_two_factor OTP secret from old format to new format'
+  task migrate_encryption_format: [:environment] do
+    # Find user records 1_000 at a time
+    User.find_each do |user|
```
We should only do it for users having an otp secret defined (2.5% of users on mastodon.social)

mjankowski replied:
Yup -- first pass I basically copied their migration script example directly. Will push update with conditions to limit to only relevant users.

mjankowski committed:
Only run for users with 2fa enabled

mjankowski commented:
I added an update rake task based on the demo migration script in the project, and modified it to only process users who have enabled OTP auth.

I did a little local test run here as well:

- Running `main` branch, create a local user with email/pass
- Enable OTP for this user, scanning QR, saving backup, etc
- Confirm signing in/out as normal requires OTP on the main branch, all good
- Stop the server, switch to this PR branch, run normal db migrations, run the migration rake script
- Watched the log, saw the value for the user be updated
- Start server, confirm sign-in flow still requires OTP (it did), and confirm the values generated by the previously stored seed still work (they do)

renchap commented:
Nice! This will probably need to be moved to a post-migrate migration, but this seems promising.

Can you try to enable OTP for a new user between the new-code deployment and the rake task, to ensure that any user who does it until the rake task it ran uses the new field?

mjankowski commented:
Sure, just did this test...

- Create a user on main, do NOT enable OTP, confirm normal email/pass login is fine
- Switch to PR branch, run db migrations but NOT migration script
- Confirm they can still sign in w/ email/pass (fine)
- Enable OTP for them, confirm sign in/out (yes)
- Run data migration script, confirm they do NOT update (since OTP created with new code)
- Confirm they can still sign in/out after data migration (yes)

mjankowski commented:
> If I understand correctly, admins will need to set the new secrets before running the new code. How will this be communicated?

Yes they would. Presumably this would be included in release notes / upgrade instructions?

If they don't, as the code stands now, the docker asset build portion would work fine (from placeholder values) but then they'll get a key not found error from the `fetch` trying to configure the value when they attempt to launch the rails process in any way.

We could potentially add a "generate some values for me and give me a string with all the env vars to copy/paste somewhere" sort of rake task that admins could run and take it's output to add to their env config? Similar to what the setup script does now, but just for AR encryption.

Separately, the setup script may need these env vars added as well?

mjankowski committed:
Convert migrate script from rake task to "post" "migrate" "migration"

mjankowski commented:
Rebased, and converted the migration script from a rake task to a "post migrate" migration.

mjankowski committed:
Values from initializer

renchap commented on Gemfile:
```diff
@@ -31,7 +31,7 @@ gem 'browser'
 gem 'chewy', '~> 7.3'
 gem 'devise', '~> 4.9'
-gem 'devise-two-factor', '~> 4.1'
+gem 'devise-two-factor'
```
We probably want to leave a version constraint here, like the other gems (maybe there is even a cop for it?)

renchap commented on config/environments/development.rb:
```diff
@@ -87,8 +87,7 @@
 
-  # We provide a default secret for the development environment here.
-  # This value should not be used in production environments!
+  # TODO: Remove once devise-two-factor data migration complete
```
We need to define when we remove this. As we support multiple version upgrades (4.1 => 4.4), should this be for Mastodon 5.0, and we ask users to first update to the latest 4 version before updating to 5.0?

(same for the other places where we have this TODO)

renchap replied:
It is not, but it would probably be a good thing to do it.

renchap reviewed (commented):
Looks good to me overall, with 2 small comments.

I would like a review from @ClearlyClaire as well

ClearlyClaire commented:
Should the rake task be a post-deployment migration instead?

mjankowski commented:
> Should the rake task be a post-deployment migration instead?

As the PR stands now, it is done as a post_deploy migration. I originally had it as a rake task, but based on renchap feedback I converted to a migration.

I'm actually not sure which approach is better, and can see arguments for both:

- The gem upgrade guide recommends a rake task - they don't explain why but I suspect it's out of a desire to slightly decouple the schema changes from the data migration portion, and maybe allow more custom/unique business logic to exist in rake tasks for anyone who has that need.
- Generally speaking I think it's best practice to attempt to change only db structure in migrations, and leave data changes as one-off rake tasks (which are then deleted after being run) ... but obviously this should be a case by case assessment.
- Given our requirement of never deleting migrations, that probably works in favor of doing a rake task for what is a one-off need that does not need to be a forever part of the codebase (after the legacy code is removed, the migration will still be present and will be doing nothing)
- The post migrate migrations will run more automatically as part of people's existing update/deploy processes, so less room for user error or forgetfulness issues there. I assume that's the general motivation for doing these things as migrations instead of tasks that need instructions added to upgrade guies?
- The rake task could be run multiple times if needed, whereas the migration will run once and then not be runnable again (without atypical db adjustments)

renchap commented:
> * The post migrate migrations will run more automatically as part of people's existing update/deploy processes, so less room for user error or forgetfulness issues there. I assume that's the general motivation for doing these things as migrations instead of tasks that need instructions added to upgrade guies?

Yes, this is the goal of having them in migrations. We need to care about upgrade processes much more than most Rails apps where there are only a few deployment of them within the same company ("internal" apps), so they can afford having more custom processes.

> * The rake task could be run multiple times if needed, whereas the migration will run once and then not be runnable again (without atypical db adjustments)

Migrations are wrapped into transactions, so if there is an error, everything is rolled back and the migration can be ran again?

mjankowski commented:
Cool, thanks for confirming re: motivations -- agreed that there's some clear upside to minimizing the number of tasks that people of different expertise levels running the app in heterogenous environments have to do.

RE: the multiple runs thing, I meant this more for the success case than the failure case. Once the migration runs that run will be recorded and it won't run again. In theory the rake task could be run over and over again (ie, to address an edge case where someone activated 2fa like mid-script run or something and was missed the first time...?)

Either way -- sounds like post_migrate is the desired state of things, which is how it stands now in the PR. Curious if there are other issues/concerns, and/or ways we could further optimize.

renchap commented:
> RE: the multiple runs thing, I meant this more for the success case than the failure case. Once the migration runs that run will be recorded and it won't run again. In theory the rake task could be run over and over again (ie, to address an edge case where someone activated 2fa like mid-script run or something and was missed the first time...?)

In theory, as this is run in a post migration, all processes are using the new code while this is running, so there should not be any issues.

I will let Claire have another look at it, I am fine with the PR as it is.

ClearlyClaire commented on db/post_migrate/20240307180905_migrate_devise_two_factor_secrets.rb:
```diff
@@ -0,0 +1,33 @@
+  def up
+    users_with_otp_enabled.find_each do |user|
+      # Gets the new value on already-updated users
+      # Falls back to legacy value on not-yet-migrated users
+      otp_secret = user.otp_secret
```
How does it fall back to the legacy value?

mjankowski replied:
It's built-in behavior in the gem -- tries to read from column first, and then falls back to reading from the legacy calculated value when empty.

ClearlyClaire replied:
But `MigrationUser` isn't using the gem, is it?

ClearlyClaire replied:
One issue is that it makes the migration script very brittle and it will likely break with future code/schema updates.

I think we can reduce the changes of that happening while keeping the maintenance burden low by keeping a `MigrationUser` stub but making it use Devise.

If the maintenance burden does become too much in the future, I guess we can make an exception for this migration and have an explicit migration breakpoint, or offer alternatives (e.g. rake task) to admins.

In any case, I think a migration test is needed so that we catch this issue and dropping the migration path is a conscious choice that we can document, instead of something breaking without us being aware.

ClearlyClaire replied:
I'll try this if you don't mind.

ClearlyClaire replied:
I've pushed a test and fixed the following issues with the post-deployment migration:
- over-reliance on `User` model causing failures in some upgrade paths
- it could overwrite post-update OTP secrets with pre-update OTP secrets if users reset OTP in the meantime

mjankowski committed:
Use User directly to include legacy otp behavior

ClearlyClaire committed:
Add migration test for OTP secret migration

ClearlyClaire committed:
Fix OTP secret post-deployment migration

mjankowski commented:
Rebased against main after the unique index PR merged (both modified tests rake); preserved the additional test from here.

mjankowski commented on db/post_migrate/20240307180905_migrate_devise_two_factor_secrets.rb:
```diff
@@ -0,0 +1,39 @@
+    devise :two_factor_authenticatable,
+           otp_secret_encryption_key: Rails.configuration.x.otp_secret
+
+    include LegacyOtpSecret # Must be after the above `devise` line in order to override the legacy method
```
One concern here -- we are going to delete that module at some point in future release.

Maybe that's fine and we just also update this migration to do nothing at that point, or inline it here?

Hanage999 commented:
[cipher.final] (https://github.com/mastodon/mastodon/blob/1e7d5d2957678788fdea8ade77eced98848ff4ff/app/models/concerns/legacy_otp_secret.rb#L75) fails with Openssl::Cipher::CipherError on an old user, which leads to a migration failure [here](https://github.com/mastodon/mastodon/blob/1e7d5d2957678788fdea8ade77eced98848ff4ff/db/post_migrate/20240307180905_migrate_devise_two_factor_secrets.rb#L21).

> OpenSSL::Cipher::CipherError: OpenSSL::Cipher::CipherError
> /home/mastodon/live/app/models/concerns/legacy_otp_secret.rb:75:in 'final'
> /home/mastodon/live/app/models/concerns/legacy_otp_secret.rb:75:in 'legacy_otp_secret'
> /home/mastodon/live/vendor/bundle/ruby/3.3.0/gems/devise-two-factor-5.0.0/lib/devise_two_factor/models/two_factor_authenticatable.rb:19:in 'otp_secret'
> /home/mastodon/live/db/post_migrate/20240307180905_migrate_devise_two_factor_secrets.rb:21:in 'block in up'

The user whose OTP caused the error registered on my server on 2017.5.19 and was somehow suspended on 2018.7.13, so they enabled OTP at some point between those two dates.

I disabled their OTP, and the migration went on successfully.

I regret I didn't dig deeper into the problem, but thought it's good to let you know anyway.

ClearlyClaire commented:
Have you ever edited `OTP_SECRETS` during the existence of your server? This may explain the issue.

I guess we could catch and ignore the error, but this would risk silently breaking TOTP for existing users. I'm not sure what the best solution would be.

Hanage999 commented:
I found out that I did change OTP_SECRETS once, sometime around 2022. 

Thank you for your insight!

Hanage999 commented:
And it also came to my mind that I changed OTP_SECRET because @Gargron recommended, well, sort of, to do so when he notified Mastodon admins via email of a vulnerability found in v2.4.0rc1-rc2, back in May, 2018.

It is possible that a server created before that could experience the same issue, if one of its dead accounts once enabled OTP and left it un-reset.

It would be a good idea to recommend that the admin of and old server should check if its dead and old accounts once enabled OTP and left it untouched.

mjankowski commented:
Hmm, we should probably do something to handle that scenario.

If OTP is already broken for those users, it's probably fine to catch the error and silently not update and move on. If it's in some way recoverable, we should attempt that in the migration and/or with doc/guidance ahead of the migration running.

Hanage999 commented:
Simply disabling OTPs of dead accounts (= those suspended, deleted or inactive for a long time) before the DB migration would be fine, maybe?

mjankowski commented:
Maybe, but I don't think that's actually relevant to the issue. The query is grabbing all users that have `otp_required_for_login` enabled and where `otp_secret` column is null (meaning they have not yet been migrated to this approach). There's nothing in there about suspended, deleted, account age, etc ... I think you just happened to hit the issue with a user like that.

It probably is more likely that "a changed `OTP_SECRET` value" might overlap with "old users on long running instances" ... just because thats probably more likely to have been changed the older the instance, etc.

ClearlyClaire commented:
Tried to address this in https://github.com/mastodon/mastodon/pull/30344
